cram_3d_world/cram_btr_spatial_relations_costmap/src/prolog.lisp

```diff
@@ -432,6 +432,63 @@
   ;;             (equal ?orientation-type :random)))
   ;;   (generate-orientations ?orientation-type ?environment-link nil ?costmap))
 
+
+  ;;;;;;;;;;;;;;; spatial relation ABOVE for environment objects ;;;;;;;;;;;;
+  ;; ABOVE works similar to ON. Major differences are
+  ;; 1. prismatic containers are opened before manipulation
+  ;; 2. it requires a z-offset to be provided to calculate height
+  (<- (costmap:desig-costmap ?designator ?costmap)
+    (and (desig:desig-prop ?designator (:above ?object))
+         (not (desig:desig-prop ?designator (:attachment ?_)))
+         (not (desig:desig-prop ?designator (:attachments ?_))))
+    (desig:desig-prop ?designator (?original-tag ?object))
+    (spec:property ?object (:urdf-name ?urdf-name))
+    (spec:property ?object (:part-of ?environment-name))
+    (once (or (and (desig:desig-prop ?designator (:z-offset ?z-offset))
+                   (lisp-pred typep ?z-offset float))
```

Comment:
- lisp-type is faster
- should use spec:property instead of desig:desig-prop (for that need to add the spec for ``z-offset``)